internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,289 @@
+	// - Adds the finalizer if not present, ensuring proper cleanup.
+	if controllerutil.AddFinalizer(sa, kargoapi.FinalizerName) {
+		if err := r.client.Update(ctx, sa); err != nil {
+			return ctrl.Result{}, err
+		}
+		logger.Debug("Added finalizer to ServiceAccount", "serviceaccount", sa.Name)
+		return ctrl.Result{}, nil
+	}
+
+	logger.Debug("Creating RoleBindings for ServiceAccount",
+		"serviceaccount", sa.Name,
+	)
+	return ctrl.Result{}, r.ensureControllerPermissionsFn(ctx, sa)
+}
+
+// ensureControllerPermissions grants the controller ServiceAccount necessary access to all Projects
+// by creating RoleBindings in each Project's namespace. This function ensures that the
+// controller has the necessary permissions to manage resources across all Projects.
+func (r *reconciler) ensureControllerPermissions(ctx context.Context, sa *corev1.ServiceAccount) error {
+
+	roleBindingName := fmt.Sprintf("%s-readonly-secrets", sa.Name)
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceaccount", sa.Name,
+		"roleBinding", roleBindingName,
+	)
+	logger.Info("starting to create RoleBindings for all Projects")
+
+	projectList := &kargoapi.ProjectList{}
+	if err := r.client.List(ctx, projectList); err != nil {
+		return fmt.Errorf("error listing Projects: %w", err)
+	}
+
+	// Loop through each Project to create or update the corresponding RoleBinding.
+	for _, project := range projectList.Items {
+
+		roleBinding := &rbacv1.RoleBinding{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      roleBindingName,
+				Namespace: project.Name,
+			},
+			RoleRef: rbacv1.RoleRef{
+				APIGroup: rbacv1.GroupName,
+				Kind:     "ClusterRole",
+				Name:     "kargo-controller-secrets-readonly",
+			},
+			Subjects: []rbacv1.Subject{
+				{
+					Kind:      "ServiceAccount",
+					Name:      sa.Name,
+					Namespace: r.cfg.KargoNamespace,
+				},
+			},
+		}
+
+		// Create a new RoleBinding if it doesn't exist, or update it if it does.
+		if err := r.client.Get(ctx, client.ObjectKey{Name: roleBindingName, Namespace: project.Name}, roleBinding); err != nil {
+			if kubeerr.IsNotFound(err) {
+				// Create RoleBinding
+				if err := r.createRoleBindingFn(ctx, roleBinding); err != nil {
+					return fmt.Errorf("error creating RoleBinding %q for ServiceAccount %q in Project namespace %q: %w",
+						roleBinding.Name, sa.Name, project.Name, err)
+				}
+				logger.Info("Created RoleBinding for ServiceAccount", "roleBinding", roleBindingName, "namespace", project.Name)
+			} else {
+				return fmt.Errorf("error retrieving RoleBinding %q in namespace %q: %w", roleBindingName, project.Name, err)
+			}
+		} else {
+			// Update RoleBinding
+			if err := r.client.Update(ctx, roleBinding); err != nil {
+				return fmt.Errorf("error updating RoleBinding %q in Project namespace %q: %w", roleBinding.Name, project.Name, err)
+			}
+			logger.Debug("Updated RoleBinding for ServiceAccount", "roleBinding", roleBindingName, "namespace", project.Name)
```

Comment:
I still believe that having bothered with the get above, you you probably check here whether updating this RB is even necessary. (Unnecessary updates just create potential for problems.)

Alternatively, if you _don't_ want to do that, then the get wasn't worth doing. You can jump straight to the create and update if you get an error saying it already exists.

Whichever of these two approaches you prefer is fine with me.